src/compiler/chainrules.jl

```diff
@@ -236,32 +260,85 @@ end
     zygote2differential(dx, primal)
 
 Convert input `dx` from the Zygote format to the ChainRules differential types.
+This is similar to `wrap_chainrules_input(dx)`, but because it gets `primal::T`,
+it can turn `NamedTuple`s into `Tangent{T}(...)` not `Tangent{Any}(...)`.
 """
 zygote2differential(x, primal) = z2d(x, primal)
 zygote2differential(::Nothing, ::Any) = NoTangent()
 zygote2differential(t::Tuple, primal::Tuple) = map(z2d, t, primal)
 zygote2differential(t::Tuple, primal) = (@warn "primal should be a tuple, not $primal"; return t)
-z2d(x, ::Any) = x
+
 z2d(::Nothing, ::Any) = NoTangent()
-z2d(a::AbstractArray{<:Number}, primal::AbstractArray{T}) where T = a
-# Could probably `reinterpret` instead of broadcasting here -- TODO
-z2d(a::AbstractArray, primal::AbstractArray{T}) where T = z2d.(a, primal)
+z2d(::Tuple{Vararg{Nothing}}, ::Tuple) = NoTangent()  # collapse all-zero case
+z2d(dx, ::Any) = dx
+z2d(dx::AbstractArray{<:Number}, primal::AbstractArray) = dx
+z2d(dx::AbstractArray{<:AbstractArray{<:Number}}, primal::AbstractArray) = dx
+z2d(dx::AbstractArray, primal::AbstractArray) = map(z2d, dx, primal)
+#=
+# As an optimisation, we can convert by `reinterpret` for bitstypes, e.g. arrays of tuples of numbers
+function z2d(dx::AbstractArray{S}, primal::AbstractArray{P}) where {S,P}
+  if isbitstype(S)
+    T = wrap_chainrules_input(S)
+    reinterpret(Tangent{P,T}, dx)
+  else
+    map(z2d, dx, primal)
+  end
+end
+=#
+
 # Note: this should never be hit if we are converting things right, but it seems to be
 # happening in the wild for sufficiently weird functions/types.
 # This fixes most (all?) cases, but it would be good to find what we miss.
 z2d(x::Union{AbstractZero, Tangent}, ::Any) = return x
-function z2d(t::Tuple, primal::Tuple)
-  tp::Tuple = map(z2d, t, primal)
-  primal_type = typeof(primal)
-  return canonicalize(Tangent{primal_type, typeof(tp)}(tp))
+
+function z2d(delta::Tuple, primal::Tuple)
+  backing = map(z2d, delta, primal)
+  if backing isa Tuple{Vararg{AbstractZero}}
+    return NoTangent()  # collapse all-zero case
```

Comment:
There seems to be a lot of "collapsing zero" going on. The accumulation step already takes care of the collapsing for the most part. Why can't we let the zeros cases get handled in the accumulation step? Is this related to zygote sometimes producing `Array{Nothing}`?

Reply:
Without it the DistributionsAD tests fail. The main reason is that ChainRules expects non-differentiable objects to have a `NoTangent` derivative but without this fix we end up with stuff like `Tangent{T}(; a = NoTangent(), b = [NoTangent(), NoTangent()], c = Tangent{S}(; d = NoTangent()))`.

Reply:
A stacktrace would be helpful.

Reply:
There are more explanations in https://github.com/TuringLang/DistributionsAD.jl/pull/203#issuecomment-950240079. Unfortunately, it seems there are no GH actions logs of the original issue since I fixed it (partially) before adding and pushing a new commit.